src/nvim/tui/input.c

```diff
@@ -368,19 +402,37 @@ static bool handle_forced_escape(TermInput *input)
   return false;
 }
 
-static void set_bg_deferred(void **argv)
+static void set_bg(void **argv)
 {
   char *bgvalue = argv[0];
   if (!option_was_set("bg") && !strequal((char *)p_bg, bgvalue)) {
     // Value differs, apply it.
     if (starting) {
       // Wait until after startup, so OptionSet is triggered.
-      do_cmdline_cmd((bgvalue[0] == 'l')
-                     ? "autocmd VimEnter * ++once ++nested set bg=light"
-                     : "autocmd VimEnter * ++once ++nested set bg=dark");
-    } else {
-      set_option_value("bg", 0L, bgvalue, 0);
-      reset_option_was_set("bg");
+      if (is_remote_client || (!headless_mode && !embedded_mode && !silent_mode)) {
+        Array args = ARRAY_DICT_INIT;
+        Dictionary opts = ARRAY_DICT_INIT;
+        Array value = ARRAY_DICT_INIT;
+
+        ADD(value, INTEGER_OBJ(0));
+        ADD(value, STRING_OBJ(cstr_as_string(xstrdup(bgvalue))));
+        ADD(value, INTEGER_OBJ(0));
+        PUT(opts, "bg", ARRAY_OBJ(value));
+        
+        ADD(args, DICTIONARY_OBJ(opts));
+        
+        if (is_remote_client) {
```

Comment:
same here, the channel id should be known at this point, just call `rpc_send_event` with it.